src/installation/guides/fde.md

```diff
@@ -201,10 +209,10 @@ configured to automatically unlock the encrypted volume on boot. First, generate
 a random key.
 
 ```
-# dd bs=512 count=4 if=/dev/urandom of=/boot/volume.key
-4+0 records in
-4+0 records out
-2048 bytes (2.0 kB, 2.0 KiB) copied, 0.000421265 s, 4.9 MB/s
+# dd bs=512 count=64 if=/dev/random of=/boot/volume.key
+64+0 records in
+64+0 records out
+32768 bytes (33 kB, 32 KiB) copied, 0.000783849 s, 41.8 MB/s
```

Comment:
Increasing key size is OK, as long as it works, but I don't know if we should recommend `/dev/random`. In a virtual machine context it might just block, and in a non virtual context it probably has enough etropy already from the hardware. https://www.2uo.de/myths-about-urandom/

Reply:
From the very source you linked:

> What about entropy running low?
> It doesn't matter.
> 
> The underlying cryptographic building blocks are designed such that an attacker cannot predict the outcome, as long as there was enough randomness (a.k.a. entropy) in the beginning. A usual lower limit for “enough” may be 256 bits. No more.
> 
> Considering that we were pretty hand-wavey about the term “entropy” in the first place, it feels right. As we saw, the kernel's random number generator cannot even precisely know the amount of entropy entering the system. Only an estimate. And whether the model that's the basis for the estimate is good enough is pretty unclear, too.

256 bits of true entropy is more than enough for sure, the problem is that the kernel is estimating it and if the hardware is atypical it can drastically overestimate it, that's why the larger key size - though it's an ugly hack.

I personally just take 30 seconds of /dev/random output and channel it into sha512sum and call it.

Using urandom will probably be ok but when it isn't (even the kernel is estimating it has no entropy)

/dev/random was changed btw, they removed the entropy subtraction nonsense. The only difference form urandom now I believe is blocking at the very beginning if there isn't enough entropy.

Reply:
Making the "file" larger doesn't make the key larger, it's still going to just read the first (by default) 512 bits.

[cryptsetup.c](https://gitlab.com/cryptsetup/cryptsetup/-/blob/master/src/cryptsetup.c#L1730)

Reply:
Really, `bs=1 count=64` would be enough.

Reply:
> I personally just take 30 seconds of /dev/random output and channel it into sha512sum and call it.

I don't understand what you mean by this. Do you take the actual sha512 (i.e. 512 bits) or the hex representation (meaning 128 'a-f0-9`)?

Reply:
The bits of course. Basically do an entropy accumulation on top of what the kernel is doing.

Reply:
Are you sure about the keyfile because afaik LUKs should be hashing the entire keyfile, you linked to dm-crypt which does not use LUKs key slots...

Reply:
I always assumed the keyfile was being hashed, if you are correct this is really poor practice.

Reply:
> you linked to dm-crypt which does not use LUKs key slots

The function I linked to is `action_luksAddKey` ...